File: java/greedy/jump_game.java
class Solution {
    public boolean canJump(int[] nums) {
        int target = nums.length-1;
        for(int i=target; i>=0; i--){
            if(i+nums[i] >= target){
                target = i;
            }
        }

        return target == 0;
    }
}

File: python/greedy/jump_game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        size = len(nums)
        cache = [-1 for _ in range(size)]
        
        def dfs(i):
            if i >= size -1:
                return True
            if cache[i] != -1:
                return True if cache[i] == 1 else False
            
            for j in range(nums[i], 0, -1):
                if dfs(i+j):
                    cache[i] = True
                    return True

            cache[i] = False
            return False


        return dfs(0)
